tutorials/decentralized-exchange-ink/1-implementation.mdx

```diff
@@ -0,0 +1,152 @@
+
+`withdraw` function is quite intuitive where we check firstly validation of the process with `check_valid_amount` to be sure that some user is not able to withdraw more money that it is previously delivered.
+
+## Errors definitions and utils
+
+Rust does not have traditional exceptions like those found in languages such as Java or Python. Instead, Rust relies on the `Result` and `Option` types for error handling and handling absence of values, respectively.
+
+```rust
+    /// Errors definitions
+    #[derive(Debug, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum Error {
+        PoolDepleted(String),
+        ZeroAmountErr(String),
+        InvalidShareErr(String),
+        ZeroLiquidityErr(String),
+        SlippageExceededErr(String),
+        InsufficientAmountErr(String),
+        NonEquivalentValueErr(String),
+        ThresholdNotReachedErr(String),
+        InsufficientLiquidityErr(String),
+    }
+```
+
+In the smart contract is implement some functions that are not strictly needed but helpful for testing, for example a place where a software engineer is able to generate tokens for tests - `faucet_brrr`. To follow K curve in an automated market maker is implemented `get_k` function. To get information about portfolio is implemented `get_information_portfolio` function and for pools information `get_pool_details`.
+
+## Additional information
+
+More information regarding ink! smart contracts programming language you will find in the [documentation](https://use.ink).
+
+It's important to note that DEX development involves complex engineering challenges and requires a deep understanding of blockchain technology, cryptography, and security practices.
+Collaboration with blockchain experts and continuous learning about evolving protocols and standards is essential in building a successful decentralized exchange.
+
+While smart contracts are designed to be immutable, there are situations where it becomes necessary to make changes or upgrades to them. Developers may encounter scenarios where they need to address critical bugs or introduce additional features.
+
+In order to build the whole smart contract use following commands:
+```bash
+git clone git@github.com:TomaszWaszczyk/ink-automated-market-maker.git
+cd ink-automated-market-maker/automated-market-maker
+cargo +nightly build --release
+```
+
+The project uses also `cargo fmt` tool in order to deliver well formatted source code, details how to use the tool you can find in the [link.](https://github.com/rust-lang/rustfmt)
+
+If you are interested in more learning and development of the smart contract one of the option is to implement `proxy pattern`. That is an example of improvement proposal that actually adds real value.
+
+## Github repository
```

Comment:
This piece of information could be good in the introduction. Something similar to: "this tutorial is based on the github repo ... Go there and explore. The tutorial will then go through explaining the key components of the code" for example.

Reply:
Fully agree. Changed that. I just want to show new people in ink! and/or Rust that here is a fully working project that is usable and compile-able from the first bits of the tutorial. There is very few working examples in the space of ink!.